### PR TITLE
MF-643 - Create end-to-end tests for openAPI

### PR DESCRIPTION
**Pull request by AryanGodara**
Signed-off-by: aryan <aryangodara03@gmail.com>

### What does this do?
Create a script to perform end-to-end testing of the openAPI
### Which issue(s) does this PR fix/relate to?
#643

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -120,3 +122,14 @@ run:
...
+
+endtoend:
+	cd scripts && . ./end-to-end.sh
+
+rundetached:
+	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
+	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
+	docker-compose -f docker/docker-compose.yml up -d
+
+composedown:
```
Comment (drasko) [inline Makefile]: I can see that this one already exists and is called `stop`

File: Makefile [hunk 2]
```diff
@@ -120,3 +122,14 @@ run:
 	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
 	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
 	docker-compose -f docker/docker-compose.yml up
+
+endtoend:
+	cd scripts && . ./end-to-end.sh
+
+rundetached:
```
Comment (drasko) [inline Makefile]: Do not add this target, rather inspect the argument given to `run` target, or pre-set ENV variable when calling make (for example `D=1 make run`) - something like [this](https://stackoverflow.com/questions/2214575/passing-arguments-to-make-run) or [this](https://chriswiegman.com/2021/08/using-parameters-in-a-make-target/)

**Review by drasko — CHANGES_REQUESTED** [review]

File: scripts/end-to-end.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#get jwt token
+JWTTOKEN=$(curl -s -S -i -X POST -H "Content-Type: application/json" http://localhost:8180/tokens -d '{"email":"'"$EMAIL"'", "password":"'"$PASSWORD"'"}' | grep -Po "token\":\"\K(.*)(?=\")")
+printf "JWT TOKEN for user is $JWTTOKEN \n"
+
+echo setting mf base path $(pwd)
+export MF_BASE_PATH=$(pwd)
+
+
+echo setting mf auth bearer token $(JWTTOKEN)
+export MF_TOKEN=$(JWTTOKEN)
```
Comment (dborovcanin) [inline scripts/end-to-end.sh]: Add an empty line at the end.

File: scripts/end-to-end.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+###
+# Runs all Mainflux microservices (builds and installs if not done already)
+# 
+# Uses Schemathesis to check the openAPI configuration with the actual endpoints
+###
+
+cd ..
+
+make dockers
```
Comment (dborovcanin) [inline scripts/end-to-end.sh]: Make dockers in dev mode, like:
`make -j $NUMPROC` and `make -j $NUMPROC dockers_dev` where `$NUMPROC` is the number of processing cores available on CI (check how to get that).

File: Makefile [hunk 5]
```diff
@@ -120,3 +122,14 @@ run:
...
+	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
+	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
+	docker-compose -f docker/docker-compose.yml up -d
+
+composedown:
+	docker-compose -f docker/docker-compose.yml down
```
Comment (dborovcanin) [inline Makefile]: Add an empty line at the end.

File: .github/workflows/st.yml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          schema: './api/openapi/websocket.yml'
+          base-url: 'http://localhost'
+          args: -H Authorization:${{ env.MF_TOKEN }}
+          # Set your token from secrets
+          token: ${{ secrets.SCHEMATHESIS_TOKEN }}
+
+
+      - name: Stop the running docker containers
+        if: always()
+        run: make composedown
```
Comment (dborovcanin) [inline .github/workflows/st.yml]: Add an empty line.

File: .github/workflows/main.yaml [hunk 7]
```diff
@@ -23,4 +23,4 @@ jobs:
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
           publish_dir: swagger-ui
-          cname: api.mainflux.io
+          cname: api.mainflux.io
```
Comment (dborovcanin) [inline .github/workflows/main.yaml]: Add empty line.

File: Makefile (hunk 2 above)
Comment (AryanGodara) [reply Makefile]: Added $(ARGS)
```
run:
	sed -i "s,file: brokers/.*.yml,file: brokers/${MF_BROKER_TYPE}.yml," docker/docker-compose.yml
	sed -i "s,MF_BROKER_URL=.*,MF_BROKER_URL=$$\{MF_$(shell echo ${MF_BROKER_TYPE} | tr 'a-z' 'A-Z')_URL\}," docker/.env
	docker-compose -f docker/docker-compose.yml up $(ARGS)
```
to run in detached: `make run ARGS="-d`
to run normally: `make run`

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: scripts/end-to-end.sh (hunk 4 above)
Comment (AryanGodara) [reply scripts/end-to-end.sh]: it has 2 cores, used lscpu command

File: .github/workflows/st.yml [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        with:
+          schema: './api/openapi/websocket.yml'
+          base-url: 'http://localhost'
+          args: -H Authorization:${{ env.MF_TOKEN }}
+          token: ${{ secrets.SCHEMATHESIS_TOKEN }}
+
+
+      - name: Stop the running docker containers
+        if: always()
+        run: make stop
```
Comment (ianmuchyri) [inline .github/workflows/st.yml]: Add empty line

File: .github/workflows/st.yaml [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+jobs:
+  api-tests:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: checkout repo
+        uses: actions/checkout@v2
+
+      - name: find number of cores
+        run: lscpu | egrep 'Model name|Socket|Thread|NUMA|CPU\(s\)'
```
Comment (drasko) [inline .github/workflows/st.yaml]: stay consistent with spaces before and after the `|`

**Review by drasko — CHANGES_REQUESTED** [review]

**Comment by dborovcanin** [comment]
API tests are added a while ago, making this pull request irrelevant.

**Pull request closed without merging**